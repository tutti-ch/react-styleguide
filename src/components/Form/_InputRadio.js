import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import { filterProps } from "../../helpers/functions";
import classes from "./Form.scss";

export default class InputRadio extends PureComponent {
  static defaultProps = {
    checked: false
  };

  static propTypes = {
    /**
     * Whether the checkbox is checked or not.
     */
    checked: PropTypes.bool,

    /**
     * The value of the input.
     */
    value: PropTypes.any,

    /**
     * The input name.
     */
    name: PropTypes.string,

    /**
     * The label for the input.
     */
    label: PropTypes.string,

    /**
     * The on change callback handler.
     */
    onChange: PropTypes.func
  };

  constructor(props) {
    super(props);

    this.state = {
      checked: props.checked
    };

    this.select = this.select.bind(this);
  }

  /**
   * Sync the value.
   */
  componentDidUpdate(oldProps) {
    if (oldProps.checked !== this.props.checked) {
      this.setState({ checked: this.props.checked });
    }
  }

  /**
   * Select the input radio.
   */
  select() {
    if (this.state.checked) {
      return;
    }

    // Toggling an input radio is not possible
    this.setState({ checked: true }, () => {
      if (typeof this.props.onChange === "function") {
        const { name, checked, value } = this.props;
        this.props.onChange(value, {
          name,
          initialValue: checked,
          formValue: value
        });
      }
    });
  }

  render() {
    const { value, checked } = this.state;
    const { label, name } = this.props;
    const props = filterProps(InputRadio.propTypes, this.props);

    return (
      // We do not use a real label here because react complains
      // about uncontrolled/controlled components.
      <span onClick={this.select} className={classes.cbLabel}>
        <input
          {...props}
          name={name}
          type="radio"
          value={value}
          checked={checked}
          onChange={this.select}
        />
        <span className={classes.labelText}>{label}</span>
      </span>
    );
  }
}
