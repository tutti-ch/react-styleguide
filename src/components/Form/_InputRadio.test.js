/* global describe, test, expect, jest */
import React from "react";
import { InputRadio } from "./_InputRadio";
import { shallow, mount } from "enzyme";

describe("(Component) InputRadio", () => {
  test("[componentWillReceiveProps] should sync the state with props", () => {
    const comp = shallow(<InputRadio />);
    const inst = comp.instance();
    inst.componentWillReceiveProps({ checked: true });
    expect(comp.state("checked")).toBe(true);
  });

  test("[select] should set the value and call the onChange from parent", () => {
    const spy = jest.fn();
    const comp = shallow(<InputRadio onChange={spy} />);
    const inst = comp.instance();
    inst.select();
    expect(comp.state("checked")).toBe(true);
    expect(spy).toHaveBeenCalled();

    comp.setProps({ onChange: undefined });
    inst.select();
    expect(spy).toHaveBeenCalledTimes(1);
    expect(comp.state("checked")).toBe(true); // A selected radio can not be unselected.
  });

  test("snapshot", () => {
    const comp = mount(
      <InputRadio value="Life is good" checked label="My label" />
    );
    expect(comp).toMatchSnapshot();
  });
});
