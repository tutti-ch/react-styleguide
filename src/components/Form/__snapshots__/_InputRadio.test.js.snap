// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) InputRadio snapshot 1`] = `
<WithWrapper
  name="input-name"
>
  <div
    className="wrapper"
  >
    <InputRadioGroup
      name="input-name"
      onChange={[Function]}
      onKeyUp={[Function]}
    >
      <div>
        <InputRadio
          checked={false}
          key=".0"
          label="My label"
          name="input-name"
          onChange={[Function]}
          value="Life is good"
        >
          <span
            className="cbLabel"
            onClick={[Function]}
          >
            <input
              checked={false}
              name="input-name"
              onChange={[Function]}
              type="radio"
            />
            <span
              className="labelText"
            >
              My label
            </span>
          </span>
        </InputRadio>
        <InputRadio
          checked={true}
          key=".1"
          label="My label"
          name="input-name"
          onChange={[Function]}
          value="Life is good"
        >
          <span
            className="cbLabel"
            onClick={[Function]}
          >
            <input
              checked={true}
              name="input-name"
              onChange={[Function]}
              type="radio"
            />
            <span
              className="labelText"
            >
              My label
            </span>
          </span>
        </InputRadio>
      </div>
    </InputRadioGroup>
  </div>
</WithWrapper>
`;
